### PR TITLE
Allow non-boxing equality for StringValues

### PR DESCRIPTION
**Pull request by benaadams**
Currently testing for equality with StringValues will box; this allows a non boxing path.

What is equal may be up for debate? 

Also see https://github.com/aspnet/HttpAbstractions/issues/384

No idea what ```Object.Equals``` is thinks currently equal though.

### CODE REVIEW

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 1]
```diff
@@ -253,6 +253,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+            switch (_values.Length)
+            {
+                case 0:
+                    return 0;
+
+                case 1:
+                    return _values[0].GetHashCode();
+
+                default:
+                    var hashCode = 0;
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: hash code combiner

Comment (benaadams) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Done

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 2]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+            switch (_values.Length)
+            {
+                case 0:
+                    return 0;
+
+                case 1:
+                    return _values[0].GetHashCode();
+
+                default:
+                    HashCodeCombiner hcc = new HashCodeCombiner();
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: needs moar `var`

Comment (benaadams) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Done

**Comment by benaadams** [comment]
Really the equalities should call each other, so ```StringValue``` should call ```string``` when is only one and ```string[]``` when more than one value; but not sure I ever get enough sleep to get that looking pretty.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 3]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
             return new StringValues(combined);
         }
 
+        public static bool Equals(StringValues objA, StringValues objB)
```
Comment (justinvp) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: The parameters should be named `left` and `right` per the framework design guidelines.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 4]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+                    return true;
+            }
+        }
+
+        public bool Equals(StringValues other)
+        {
+            return Equals(this, other);
+        }
+
+        public static bool Equals(string objB, StringValues objA)
```
Comment (justinvp) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: The parameters should be named `left` and `right` per the framework design guidelines.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 5]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+        {
+            return Equals(objA, objB);
+        }
+
+        public static bool Equals(StringValues objA, string objB)
```
Comment (justinvp) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: The parameters should be named `left` and `right` per the framework design guidelines.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 6]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+                    return false;
+            }
+        }
+
+        public bool Equals(string other)
+        {
+            return Equals(this, other);
+        }
+
+        public static bool Equals(string[] objB, StringValues objA)
```
Comment (justinvp) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: The parameters should be named `left` and `right` per the framework design guidelines.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 7]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+        {
+            return Equals(objA, objB);
+        }
+
+        public static bool Equals(StringValues objA, string[] objB)
```
Comment (justinvp) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: The parameters should be named `left` and `right` per the framework design guidelines.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 8]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+        {
+            return right.Equals(left);
+        }
+
+        public static bool operator !=(string[] left, StringValues right)
+        {
+            return !right.Equals(left);
+        }
+
+        public override bool Equals(object o)
```
Comment (justinvp) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: The parameter is named `obj` on `object.Equals`. Should match here.

File: src/Microsoft.Extensions.HashCodeCombiner.Sources/HashCodeCombiner.cs [hunk 9]
```diff
@@ -70,6 +70,7 @@ public void Add(object o)
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Add<TValue>(TValue value, IEqualityComparer<TValue> comparer)
+            where TValue : class
```
Comment (rynowak) [inline src/Microsoft.Extensions.HashCodeCombiner.Sources/HashCodeCombiner.cs]: Why is this needed?

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 10]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
         }
 
+        public static bool Equals(StringValues left, StringValues right)
+        {
+            if (IsNullOrEmpty(left))
+            {
+                return IsNullOrEmpty(right);
+            }
+
+            if (left._values == null)
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: It seems like this could be dramatically simplified by using the counts + indexer - any reason why you didn't do it this way?

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 11]
```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+                return left._value == right;
+            }
+
+            switch (left._values.Length)
+            {
+                case 1:
+                    return left._values[0] == right;
+
+                default:
+                    return false;
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Any reason with this + the `string[]` version not to construct a new `StringValues` and call that version of `.Equals`?

**Comment by rynowak** [comment]
Needs unit tests :watch:

File: src/Microsoft.Extensions.HashCodeCombiner.Sources/HashCodeCombiner.cs (hunk 9 above)
Comment (benaadams) [inline src/Microsoft.Extensions.HashCodeCombiner.Sources/HashCodeCombiner.cs]: Hmm, you're right. Was because the null equality test made no sense for a struct; but then it doesn't really matter either; so will revert.

**Comment by benaadams** [comment]
> Needs unit tests :watch:

That will certainly spur me on to simplify the execution paths :grin:

File: src/Microsoft.Extensions.Primitives/StringValues.cs (hunk 10 above)
Comment (benaadams) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Lack of sleep... added a commit following your suggestion that should be far more sensible?

**Comment by benaadams** [comment]
Changed to much simpler compare at @rynowak's suggestion

**Comment by benaadams** [comment]
@rynowak added tests

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 12]
```diff
@@ -253,6 +254,157 @@ public static StringValues Concat(StringValues values1, StringValues values2)
             return new StringValues(combined);
         }
 
+        public static bool Equals(StringValues left, StringValues right)
+        {
+            // .Count is an expression not a numeric so cache locally
+            var count = left.Count;
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Do you have any perf data that suggests this is important?

Side-note, if you didn't leave the comment, I'd just assume you were being tidy.

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 13]
```diff
@@ -253,6 +254,157 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+                case 1:
+                    return _values[0].GetHashCode();
+
+                default:
+                    var hcc = new HashCodeCombiner();
+                    for (var i = 0; i < _values.Length; i++)
+                    {
+                        hcc.Add(_values[i].GetHashCode());
+                    }
+                    return hcc.CombinedHash;
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: If you want to do something simpler here you can just use the `default` case and delete the others

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 14]
```diff
@@ -253,6 +254,157 @@ public static StringValues Concat(StringValues values1, StringValues values2)
...
+                    return 0;
+
+                case 1:
+                    return _values[0].GetHashCode();
+
+                default:
+                    var hcc = new HashCodeCombiner();
+                    for (var i = 0; i < _values.Length; i++)
+                    {
+                        hcc.Add(_values[i].GetHashCode());
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: You don't need to call `GetHashCode` here, the combiner does that for you - this is going to hash the string and then hash the hash

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs [hunk 15]
```diff
@@ -302,5 +302,55 @@ public void Concat(StringValues stringValues, string[] array)
...
+
+            aString = "abc";
+            string bString = "bcd";
+            aStringArray = new[] { aString, bString };
+            stringValues = aStringArray;
+            Assert.Equal(2, stringValues.Count);
+
+            Assert.True(StringValues.Equals(aStringArray, stringValues));
+            Assert.True(StringValues.Equals(new StringValues(aStringArray), stringValues));
+            Assert.False(StringValues.Equals(new StringValues(new[] { bString, aString }), stringValues));
```
Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: This whole thing should be a data driven test instead of one big blob. Use `[Theory]`

**Comment by rynowak** [comment]
much simpler :watch:

File: src/Microsoft.Extensions.Primitives/StringValues.cs (hunk 12 above)
Comment (benaadams) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Might not be but as ```Count => _values?.Length ?? (_value != null ? 1 : 0)``` I didn't want to use it as the test condition on the following for loop. However unlikely to be many values in StringValues so can take it or leave it?

Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Sure, seems fine as is.

**Comment by benaadams** [comment]
Theory and a bunch more tests, including negatives.

Had to add a skip as `(new StringValues(null)).ToString() == ""` rather than `null`
And expanding the data to include 0 length arrays doesn't throw exceptions for `CopyTo` so added skip for that.

**Comment by rynowak** [comment]
> Had to add a skip as (new StringValues(null)).ToString() == "" rather than null

`ToString()` generally never returns null

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs [hunk 16]
```diff
@@ -263,13 +292,15 @@ public void Contains(StringValues stringValues, string[] expected)
...
-
             string[] actual = new string[expected.Length];
-            Assert.Throws<ArgumentOutOfRangeException>(() => collection.CopyTo(actual, -1));
-            Assert.Throws<ArgumentException>(() => collection.CopyTo(actual, actual.Length + 1));
+
+            if (collection.Count > 0)
+            {
+                Assert.Throws<ArgumentException>(() => collection.CopyTo(tooSmall, 0));
+                Assert.Throws<ArgumentOutOfRangeException>(() => collection.CopyTo(actual, -1));
+                Assert.Throws<ArgumentException>(() => collection.CopyTo(actual, actual.Length + 1));
```
Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: These should be separate tests.

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs [hunk 17]
```diff
@@ -302,5 +333,81 @@ public void Concat(StringValues stringValues, string[] array)
...
+            Assert.True(StringValues.Equals(stringValues, new StringValues(expected)));
+            Assert.True(StringValues.Equals(new StringValues(expected), stringValues));
+
+            Assert.True(StringValues.Equals(stringValues.ToArray(), new StringValues(expected)));
+            Assert.True(StringValues.Equals(new StringValues(expected), stringValues.ToArray()));
+            Assert.True(StringValues.Equals(new StringValues(stringValues.ToArray()), new StringValues(expected)));
+            Assert.True(StringValues.Equals(new StringValues(expected), new StringValues(stringValues.ToArray())));
+
+            // StringValues.ToString converts null to "" so skip for null values
+            if (expected != null)
```
Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: Yeah this is totally expected

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs (hunk 16 above)
Comment (benaadams) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: Separate sections or actually separate tests? (as just gathered them for the if test)

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs [hunk 18]
```diff
@@ -302,5 +333,81 @@ public void Concat(StringValues stringValues, string[] array)
...
+
+            Assert.False(StringValues.Equals(stringValues.ToArray(), new StringValues(notEqual)));
+            Assert.False(StringValues.Equals(new StringValues(notEqual), stringValues.ToArray()));
+            Assert.False(StringValues.Equals(new StringValues(stringValues.ToArray()), new StringValues(notEqual)));
+            Assert.False(StringValues.Equals(new StringValues(notEqual), new StringValues(stringValues.ToArray())));
+
+            Assert.False(StringValues.Equals(stringValues.ToString(), new StringValues(notEqual)));
+            Assert.False(StringValues.Equals(new StringValues(notEqual), stringValues.ToString()));
+            Assert.False(StringValues.Equals(new StringValues(stringValues.ToString()), new StringValues(notEqual)));
+            Assert.False(StringValues.Equals(new StringValues(notEqual), new StringValues(stringValues.ToString())));
```
Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: Why is there so much stuff here? I'd expect to see a few simple cases calling each of the equality methods that you added.

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs (hunk 16 above)
Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: https://github.com/aspnet/Home/wiki/Engineering-guidelines#unit-test-structure

We don't have a concept like "sections" - they should be separate tests.

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs (hunk 18 above)
Comment (benaadams) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: To cover all the implicit conversion paths between ```string```/```string[]```/```StringValue```, with parameter left, parameter right etc Trim them down and call each specific method? 

Should still be 23 different calls for that

Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: I must be missing something, why are the tests for `.Equals` also responsible for testing all of the conversions and all permutations of the conversions?

Comment (benaadams) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: Removed a bunch of the tests, but also added a few more e.g. for `operator==`, `object` and instance that weren't being tested.

**Comment by benaadams** [comment]
Hmm... should probably change the tests to a `[Fact]` test for all the methods, to make sure they resolve to the single `Equals` method; then do the `[Theory]` test on that one method, rather than calling them again and again...?

**Comment by benaadams** [comment]
@rynowak slightly saner number of tests?

Though CI is complaining about dependencies :cry: 
>    Unable to locate Dependency Microsoft.Extensions.DependencyInjection.Abstractions >= 1.0.0-rc1-15722
    Unable to locate Dependency Microsoft.Extensions.Logging.Abstractions >= 1.0.0-rc1-15722

**Comment by benaadams** [comment]
Sanitised tests to sensible number; while covering the different functions.

File: test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs [hunk 19]
```diff
@@ -302,5 +333,97 @@ public void Concat(StringValues stringValues, string[] array)
...
+            Assert.True(equalString == equalStringValues);
+            Assert.True(equalStringValues == equalStringArray);
+            Assert.True(equalStringArray == equalStringValues);
+
+            Assert.False(equalStringValues != equalStringValues);
+            Assert.True(equalStringValues != notEqualStringValues);
+            Assert.True(equalStringValues != notEqualString);
+            Assert.True(notEqualString != equalStringValues);
+            Assert.True(equalStringValues != notEqualStringArray);
+            Assert.True(notEqualStringArray != equalStringValues);
```
Comment (rynowak) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: This is looking better overall. The inputs to this and `Equals_Instance` and `Equals_Object` should be using `[Theory]`

`==` and `!=` should be separate tests using the same inputs but the opposite assert

Comment (benaadams) [inline test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs]: Not sure about `Equals_Instance` as they are all different data types? 

Putting the other changes in can using less hardcoded strings.

**Comment by rynowak** [comment]
:shipit: - merging this
